gradle-plugin/plugin/build.gradle.kts

```diff
@@ -63,34 +56,35 @@ val supportedKotlinVersions =
   setOf(
     "1.8.21",
   )
-
-// Creates a classpath file for each version of AGP we wish to test against. This allows the functional tests to target
-// a particular version using the GradleRunner.
-supportedAgpVersions.forEach { version ->
-  val configuration = configurations.create("agp$version")
-  dependencies.add(configuration.name, "com.android.tools.build:gradle:$version")
-  val classpathFile =
-    agpClasspathDir.resolve("agp-classpath-$version.txt").apply {
-      ensureParentDirsCreated()
+val writeSupportedVersionsFiles = tasks.register("writeSupportedVersionsFiles") {
+  inputs.property("supportedAgpVersions", supportedAgpVersions)
+  inputs.property("supportedKotlinVersion", supportedKotlinVersions)
+  outputs.dirs(agpClasspathDir, kgpClasspathDir)
+  // Creates a classpath file for each version of AGP we wish to test against. This allows the
+  // functional tests to target a particular version using the GradleRunner.
+  doLast {
+    supportedAgpVersions.forEach { version ->
+      val configuration = configurations.create("agp$version")
+      dependencies.add(configuration.name, "com.android.tools.build:gradle:$version")
+      val classpathFile =
+        agpClasspathDir.get().asFile.resolve("agp-classpath-$version.txt")
+      classpathFile.writeText(configuration.joinToString(separator = "\n"))
     }
-  classpathFile.writeText(configuration.joinToString(separator = "\n"))
-}
 
-supportedKotlinVersions.forEach { version ->
-  val configuration = configurations.create("kgp$version")
-  dependencies.add(configuration.name, "org.jetbrains.kotlin:kotlin-gradle-plugin:$version")
-  val classpathFile =
-    kgpClasspathDir.resolve("kgp-classpath-$version.txt").apply {
-      ensureParentDirsCreated()
+    supportedKotlinVersions.forEach { version ->
+      val configuration = configurations.create("kgp$version")
+      dependencies.add(configuration.name, "org.jetbrains.kotlin:kotlin-gradle-plugin:$version")
+      val classpathFile =
+        kgpClasspathDir.get().asFile.resolve("kgp-classpath-$version.txt")
+      classpathFile.writeText(configuration.joinToString(separator = "\n"))
     }
-  classpathFile.writeText(configuration.joinToString(separator = "\n"))
+  }
 }
 
 val functionalTest: SourceSet by sourceSets.creating {
   // The classpath files are added as resources so that the test code can access them.
   resources {
-    srcDir(agpClasspathDir)
-    srcDir(kgpClasspathDir)
+    srcDir(writeSupportedVersionsFiles)
```

Comment:
this is a neat gradle trick: you can specify a task here instead of a directory and that task’s outputs are automatically wired up to this source set along with a task dependency.

Reply:
Very nice!